walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,178 @@
+      unavailableTimesOptional = new ArrayList<>();
+    }
+
+    return getAvailableTimes(request.getDuration(), unavailableTimes, unavailableTimesOptional);
+
   }
+
+  /**
+   * Get all unavailable times in which meeting request attendees are busy.
+   * 
+   * @param events to get unavailabilities from
+   * @param requestAttendees to find overlap for
+   * @return list of time ranges that are already occupied
+   */
+  public List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> requestAttendees) {
+    List<TimeRange> output = new ArrayList<>();
+
+    for (Event e : events) {
+      Set<String> eventAttendees = e.getAttendees();
+      
+      for (String eventAttendee : eventAttendees) {
+
+        // Checks to see if event has overlap with the attendees of the request
+        if (requestAttendees.contains(eventAttendee)) {
+          output.add(e.getWhen());
+
+          break; // Only one attendee is required for the event time to be unavailable. 
+        }
+      }
+    }
+    
+    return output;
+  }
+
+  /**
+   * Gets all available times for the meeting request attendees that contain the most available attendees.
+   *
+   * @param duration of meeting request
+   * @param unavailableTimes of mandatory attendees
+   * @param unavailableTimesOptional of optional attendees
+   * @return list of available time ranges sorted by start time
+   */
+  private List<TimeRange> getAvailableTimes(long duration, List<TimeRange> unavailableTimes, List<TimeRange> unavailableTimesOptional) {
+    List<TimeRange> output = new ArrayList<>();
+
+    // Finds all timeslots in which mandatory attendees can attend
+    List<TimeRange> attendeeTimes = getNonOverlappingTimes(duration, mergeOverlappingTimes(unavailableTimes));
+
+    // If there are no mandatory attendees or if there are no optional attendees, return attendeeTimes
+    if (unavailableTimesOptional.isEmpty()) {
+      Collections.sort(attendeeTimes, TimeRange.ORDER_BY_START);
+      return attendeeTimes;
+    }
+
+    // Find all timeslots in which optional attendees can attend
+    List<TimeRange> attendeeTimesOptional = getNonOverlappingTimes(duration, mergeOverlappingTimes(unavailableTimesOptional));
+
+    if (attendeeTimesOptional.isEmpty()) {
+      return attendeeTimes;
+    } else {
+      // Find overlapping times between mandatory and optional times
+      List<TimeRange> allUnavailableTimes = new ArrayList<>();
+      allUnavailableTimes.addAll(unavailableTimes);
+      allUnavailableTimes.addAll(unavailableTimesOptional);
+      Collections.sort(allUnavailableTimes, TimeRange.ORDER_BY_START);
+      
+      List<TimeRange> allAvailableTimesMerged = getNonOverlappingTimes(duration, mergeOverlappingTimes(allUnavailableTimes));
+      
+      // If none exist, return mandatory attendee times.
+      if (allAvailableTimesMerged.isEmpty()) {
+        Collections.sort(attendeeTimes, TimeRange.ORDER_BY_START);
+        return attendeeTimes;
+      } else {
```

Comment:
same thing here. Don't need to wrap this in an Else.

Reply:
Fixed it in [new PR](https://github.com/caroljli/portfolio/pull/19)!! @aditij113